### PR TITLE
Unciv server

### PR DESCRIPTION
**Pull request by yairm210**
I think I'm not extracting the contents on home save properly in the server, hence why the files returned are empty
To use: run UncivServer and UncivDesktop together, options-multiplayer, copy ip and paste ip to "server ip"

UI of options looks bad, as can be expected

### CODE REVIEW

File: desktop/src/com/unciv/app/desktop/UncivServer.kt [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+fun main() {
+    val fileFolderName = "MultiplayerFiles"
+    embeddedServer(Netty, port = 8080) {
+        routing {
+            get("/isalive") {
+                call.respondText("true")
+            }
+            post("/files/{fileName}"){
+                val fileName = call.parameters["fileName"] ?: throw Exception("No fileName!")
+                val recievedText = call.request.receiveChannel().toInputStream().toString()
```
Comment (yairm210) [inline desktop/src/com/unciv/app/desktop/UncivServer.kt:24]: This line is probably what's causing the problems, you probably need to read it too a buffer or idk

**Comment by SomeTroglodyte** [comment]
Umm, "http POST" and the concept "upload" are not the same... A POST is simply put just a GET with parameters moved from the URL to request headers. ?right?  [this](https://ryanharrison.co.uk/2018/09/20/ktor-file-upload-download.html) looks plausible...

**Comment by yairm210** [comment]
I'm a believer in http verbs having meaning, but in any case this should be a post, since the data is passed by content and not by url
And no, not exactly. Both verbs can accept headers, but headers have a limited size. For large data transfer you need content.

**Comment by SomeTroglodyte** [comment]
I still believe I'm right. POST is like an API call, classically created in conjunction with Client-side html FORM where fields under that form are automatically included in the request, but not via the URL. The URL is the server-side entity that will process said data, e.g. a php/aspx/whatever script updating a database with contact information from that form. If you use POST for uploads, the semantics must be akin to said database example, e.g. the URL is "/fileservice", there's a "gameInfo" field in the POST parameters, then you save that in the filesystem instead of a database - or in sql after all, the client won't know. As I said, service/rpc semantics. Routing "files/<filename>" to your service is strange, as there may be no actual entity at that URL... The semantics you are using are more like those of the PUT verb.

Nevertheless, since routing has the last word, the construct should still work.

maybe `val receivedText = call.receiveText()` as the dox say "To access the raw bits of the payload, you can use receiveChannel, but it is directly part of the call instead of call.request"

**Comment by SomeTroglodyte** [comment]
Congrats :party:, looks like going in the right direction. I assume code cleanup (println, package locality) is planned. You didn't adopt the coroutine approach to the server side, which is OK as the examples yield every 4MB and our files are small -> Idea turn on gzip for them to further move load away from the server?

**Comment by yairm210** [comment]
All files send to/from devices are gzipped anyway to minimize network load ;)

**Comment by yairm210** [comment]
This is a complex decision, but I think it's actually better to leave the server as part of the *general* jar rather than separate it to a different jar.

Pros:
Our current production stream gets it out to all clients (git, itch,io, flatpak) and changing all of them to include a new jar will make things both confusing to 'basic' players

Cons:
Larger jar size - if this was APK size then I'd have an issue with it, but honestly desktops have so much space they don't know what to do with it
Can't specify is as main class so it's not just `java -jar UncivServer.jar` - this is a real point, but honestly I feel that if someone is active enough to want to run a server he can read the correct command anyway.

**Comment by yairm210** [comment]
In the future, @GGGuenni can use this as a base to implement many of the points of #6360, thus giving people a more selfish reason to migrate away from Dropbox. Want a multiplayer lobby, sorry pal that's only for private servers not for Dropbox ;)

**Comment by SomeTroglodyte** [comment]
Agree 100%, zero-delay w/ pro/cons. Maybe one day make it runnable from the Options menu, even on Droid? Yea, joking.

**Comment by xlenstra** [comment]
The only alternative I could think of that would not confuse players while also not having those cons, would be to have two jar's. These should then either distribute these independently, e.g., two different games on itch, or the server should only be downloadable after following an in-game button that redirects you to a download & instruction page on our github, thus not requiring additional distribution work on our part.

Though having them in the same jar is probably fine.

**Comment by touhidurrr** [comment]
> I'm a believer in http verbs having meaning, but in any case this should be a post, since the data is passed by content and not by url And no, not exactly. Both verbs can accept headers, but headers have a limited size. For large data transfer you need content.

Normally in RESTful API's they does it like this:

`GET` get one data
`POST` create one data
`PATCH` edit one data
`PUT` create / edit multiple data (if data doesn’t exist then create new otherwise edit)
`DELETE` delete one data

**Comment by touhidurrr** [comment]
> All files send to/from devices are gzipped anyway to minimize network load ;)

The Gzipping feels a little weird to me. Cause we gzip and then base64 it. The base64 part feels unnecessary if you implementing a Gzip library anyways.

**Comment by SomeTroglodyte** [comment]
> base64

Methinks.... Because that is also used for issue communication right here - those copy to clipboard buttons in the game. You are right this need not be the format on-disk or over-the-wire. Feel up to designing an enum FileCompressionLevel {ClearText, GZipBase64, GZipBinary}, teaching it to GameSaver and MapSaver (you'll find not everything relevant is properly encapsulated in those), etc etc...?

Besides, the main source of bloat is long field names being repeated verrrrrrrryyy often in json, so the zipping part is exceptionally efficient and the 4:3 penalty of Base64 is far less than that.

**Pull request merged**